src/Symfony/Component/HttpFoundation/ParameterBag.php

```diff
@@ -105,40 +105,73 @@ public function remove(string $key)
      */
     public function getAlpha(string $key, string $default = ''): string
     {
-        return preg_replace('/[^[:alpha:]]/', '', $this->get($key, $default));
+        return preg_replace('/[^[:alpha:]]/', '', $this->getString($key, $default));
     }
 
     /**
      * Returns the alphabetic characters and digits of the parameter value.
      */
     public function getAlnum(string $key, string $default = ''): string
     {
-        return preg_replace('/[^[:alnum:]]/', '', $this->get($key, $default));
+        return preg_replace('/[^[:alnum:]]/', '', $this->getString($key, $default));
     }
 
     /**
      * Returns the digits of the parameter value.
      */
     public function getDigits(string $key, string $default = ''): string
     {
-        // we need to remove - and + because they're allowed in the filter
-        return str_replace(['-', '+'], '', $this->filter($key, $default, \FILTER_SANITIZE_NUMBER_INT));
+        return preg_replace('/[^[:digit:]]/', '', $this->getString($key, $default));
     }
 
     /**
      * Returns the parameter value converted to integer.
+     *
+     * @param bool $failOnInvalid Throw an exception when the value is not numeric
      */
-    public function getInt(string $key, int $default = 0): int
+    public function getInt(string $key, int $default = 0, /* bool $failOnInvalid = false */): int
     {
-        return (int) $this->get($key, $default);
+        $failOnInvalid = \func_num_args() >= 3 ? func_get_arg(2) : false;
+
+        if (!$failOnInvalid) {
+            trigger_deprecation('symfony/http-foundation', '6.3', 'Pass true as 3rd argument to "%s()" to throw an exception when a non-numeric value is passed to the request for parameter "%s".', __METHOD__, $key);
+
+            return (int) $this->get($key, $default);
+        }
+
+        $value = $this->filter($key, $default, \FILTER_VALIDATE_INT, ['flags' => \FILTER_REQUIRE_SCALAR]);
+        if (false === $value) {
+            throw new \UnexpectedValueException(sprintf('Parameter value "%s" cannot be converted to "int".', $key));
+        }
+
+        return $value;
+    }
+
+    /**
+     * Returns the parameter as string.
+     */
+    public function getString(string $key, string $default = ''): string
+    {
+        $value = $this->get($key, $default);
+        if (!\is_scalar($value) && !$value instanceof \Stringable) {
+            throw new \UnexpectedValueException(sprintf('Parameter value "%s" cannot be converted to "string".', $key));
+        }
+
+        return (string) $value;
     }
 
     /**
      * Returns the parameter value converted to boolean.
      */
     public function getBoolean(string $key, bool $default = false): bool
     {
-        return $this->filter($key, $default, \FILTER_VALIDATE_BOOL);
+        $value = $this->filter($key, $default, \FILTER_VALIDATE_BOOL);
```

Comment:
we need to use FILTER_NULL_ON_FAILURE if we want this to return a non-bool when an invalid value is submitted